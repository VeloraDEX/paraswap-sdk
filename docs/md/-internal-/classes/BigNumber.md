[**@paraswap/sdk**](../../README.md) • **Docs**

***

[@paraswap/sdk](../../globals.md) / [\<internal\>](../README.md) / BigNumber

# Class: BigNumber

## Implements

- [`Hexable`](../interfaces/Hexable.md)

## Constructors

### new BigNumber()

> **new BigNumber**(`constructorGuard`, `hex`): [`BigNumber`](BigNumber.md)

#### Parameters

• **constructorGuard**: `any`

• **hex**: `string`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:7

## Properties

### \_hex

> `readonly` **\_hex**: `string`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:5

***

### \_isBigNumber

> `readonly` **\_isBigNumber**: `boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:6

## Methods

### abs()

> **abs**(): [`BigNumber`](BigNumber.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:10

***

### add()

> **add**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:11

***

### and()

> **and**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:17

***

### div()

> **div**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:13

***

### eq()

> **eq**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:23

***

### fromTwos()

> **fromTwos**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:8

***

### gt()

> **gt**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:26

***

### gte()

> **gte**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:27

***

### isNegative()

> **isNegative**(): `boolean`

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:28

***

### isZero()

> **isZero**(): `boolean`

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:29

***

### lt()

> **lt**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:24

***

### lte()

> **lte**(`other`): `boolean`

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

`boolean`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:25

***

### mask()

> **mask**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:20

***

### mod()

> **mod**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:15

***

### mul()

> **mul**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:14

***

### or()

> **or**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:18

***

### pow()

> **pow**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:16

***

### shl()

> **shl**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:21

***

### shr()

> **shr**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:22

***

### sub()

> **sub**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:12

***

### toBigInt()

> **toBigInt**(): `bigint`

#### Returns

`bigint`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:31

***

### toHexString()

> **toHexString**(): `string`

#### Returns

`string`

#### Implementation of

[`Hexable`](../interfaces/Hexable.md).[`toHexString`](../interfaces/Hexable.md#tohexstring)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:33

***

### toJSON()

> **toJSON**(`key`?): `any`

#### Parameters

• **key?**: `string`

#### Returns

`any`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:34

***

### toNumber()

> **toNumber**(): `number`

#### Returns

`number`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:30

***

### toString()

> **toString**(): `string`

Returns a string representation of an object.

#### Returns

`string`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:32

***

### toTwos()

> **toTwos**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `number`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:9

***

### xor()

> **xor**(`other`): [`BigNumber`](BigNumber.md)

#### Parameters

• **other**: [`BigNumberish`](../type-aliases/BigNumberish.md)

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:19

***

### from()

> `static` **from**(`value`): [`BigNumber`](BigNumber.md)

#### Parameters

• **value**: `any`

#### Returns

[`BigNumber`](BigNumber.md)

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:35

***

### isBigNumber()

> `static` **isBigNumber**(`value`): `value is BigNumber`

#### Parameters

• **value**: `any`

#### Returns

`value is BigNumber`

#### Defined in

node\_modules/@ethersproject/bignumber/lib/bignumber.d.ts:36
