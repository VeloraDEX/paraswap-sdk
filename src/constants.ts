import { EnumerateLiteral, APIVersion } from './types';

import type { ContractMethod } from '@paraswap/core';

export {
  SwapSide,
  ContractMethod,
  ContractMethodV5,
  ContractMethodV6,
} from '@paraswap/core';

// allows to pass by name instead of importing enum first
export type ContractMethodByName = EnumerateLiteral<typeof ContractMethod>;

export const API_URL = 'https://api.paraswap.io';
export const DEFAULT_VERSION = '6.2' satisfies APIVersion;
