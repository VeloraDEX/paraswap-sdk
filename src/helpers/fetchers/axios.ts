import type { ExtraFetchParams, FetcherFunction } from '../../types';
import type { AxiosStatic } from 'axios';
import { FetcherError } from '../misc';

export type AxiosRequirement = Pick<AxiosStatic, 'request' | 'isAxiosError'>;

export const constructFetcher =
  (axios: AxiosRequirement, extra?: ExtraFetchParams): FetcherFunction =>
  async (params) => {
    try {
      const { requestParams, ...rest } = params;
      // adding apiKey to headers if it's provided
      const headers = extra?.apiKey
        ? {
            'X-API-KEY': extra.apiKey,
            ...rest.headers,
            ...requestParams?.headers,
          }
        : { ...rest.headers, ...requestParams };

      const allParams = { ...rest, ...requestParams, headers };

      const { data } = await axios.request(allParams);

      return data;
    } catch (error: any) {
      if (axios.isAxiosError(error)) {
        // attach a more meaningful message and other props
        throw new FetcherError(error);
      }

      throw error;
    }
  };
