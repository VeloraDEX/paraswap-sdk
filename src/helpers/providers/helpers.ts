import { assert } from 'ts-essentials';
import { TypedDataField } from '../../methods/common/orders/buildOrderData';

// regex from @ethersproject/hash TypedDataEncoder.constructor
// may be overly strict, but reliable
const baseTypeRegex = /^([^\x5b]*)(\x5b|$)/;

export function findPrimaryType(
  types: Record<string, TypedDataField[]>
): string {
  const candidates = Object.keys(types);
  const candidatesSet = new Set(candidates);

  candidates.forEach((candidate) => {
    const typedDataFields = types[candidate];
    if (!typedDataFields) return;

    typedDataFields.forEach(({ type }) => {
      // Get the base type (drop any array specifiers)
      const baseType = type.match(baseTypeRegex)?.[1];
      if (!baseType) return;

      // if type was referred to as a child of another type, it can't be the primaryType
      candidatesSet.delete(baseType);
    });
  });

  const [primaryType] = Array.from(candidatesSet);

  assert(
    primaryType,
    `No primary type found in SignableTypedData types, ${JSON.stringify(types)}`
  );

  return primaryType;
}
