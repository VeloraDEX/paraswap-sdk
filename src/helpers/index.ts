import { EthersV5ProviderDeps } from './providers/ethers';
import { EthersV6ProviderDeps } from './providers/ethersV6';
export { constructFetcher as constructAxiosFetcher } from './fetchers/axios';
export { constructFetcher as constructFetchFetcher } from './fetchers/fetch';
export {
  constructEthersV5ContractCaller as constructEthersContractCaller,
  constructEthersV5ContractCaller,
  EthersV5ProviderDeps,
} from './providers/ethers';
export {
  constructContractCaller as constructEthersV6ContractCaller,
  EthersV6ProviderDeps,
} from './providers/ethersV6';
export type EthersProviderDeps = EthersV5ProviderDeps | EthersV6ProviderDeps;

export {
  constructContractCaller as constructWeb3ContractCaller,
  Web3UnpromiEvent,
} from './providers/web3';
export {
  constructContractCaller as constructViemContractCaller,
  MinViemClient,
  txParamsToViemTxParams,
} from './providers/viem';
export { constructToken } from './token';
export { isFetcherError } from './misc';
export type { FetcherError, FetcherErrorInterface } from './misc';
