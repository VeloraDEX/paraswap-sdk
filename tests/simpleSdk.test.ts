import * as dotenv from 'dotenv';
import Web3 from 'web3';
import { BigNumber as BigNumberEthers, ethers } from 'ethersV5';
import { ethers as ethersV6 } from 'ethers';
import axios from 'axios';
import fetch from 'isomorphic-unfetch';
import { isAllowance, SwapSide, SimpleFetchSDK } from '../src';
import BigNumber from 'bignumber.js';

import erc20abi from './abi/ERC20.json';

import { assert } from 'ts-essentials';

import { constructSimpleSDK, SimpleSDK } from '../src/sdk/simple';
import { HardhatProvider, setupFork } from './helpers/hardhat';

dotenv.config();

jest.setTimeout(30 * 1000);

const ETH = '0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee';
const DAI = '0x6B175474E89094C44Da98b954EedeAC495271d0F';
// const HEX = '0x2b591e99afe9f32eaa6214f7b7629768c40eeb39';

const DUMMY_ADDRESS_FOR_TESTING_ALLOWANCES =
  '0xb9A079479A7b0F4E7F398F7ED3946bE6d9a40E79';

const chainId = 1;
const srcToken = ETH;
const destToken = DAI;
const srcAmount = (1 * 1e18).toString(); //The source amount multiplied by its decimals

const referrer = 'sdk-test';
const TEST_MNEMONIC =
  'radar blur cabbage chef fix engine embark joy scheme fiction master release';
//0xaC39b311DCEb2A4b2f5d8461c1cdaF756F4F7Ae9
const wallet = ethers.Wallet.fromMnemonic(TEST_MNEMONIC);
const walletV6 = ethersV6.HDNodeWallet.fromPhrase(TEST_MNEMONIC);

const web3provider = new Web3(HardhatProvider as any);

const ethersProvider = new ethers.providers.Web3Provider(
  HardhatProvider as any
);
const ethersV6Provider = new ethersV6.BrowserProvider(HardhatProvider);
const signerV6 = walletV6.connect(ethersV6Provider);

const signer = wallet.connect(ethersProvider);
const senderAddress = signer.address;

describe.each([
  ['fetch', { fetch }],
  ['axios', { axios }],
])('ParaSwap SDK: fetcher made with: %s', (testName, fetcherOptions) => {
  let paraSwap: SimpleFetchSDK;

  beforeAll(async () => {
    await setupFork({ accounts: [{ address: senderAddress, balance: 8e18 }] });

    paraSwap = constructSimpleSDK({ chainId, ...fetcherOptions, version: '5' });
  });
  test('getBalance', async () => {
    try {
      const balance = await paraSwap.swap.getBalance(senderAddress, ETH);
      expect(balance).toBeDefined();
    } catch (error: any) {
      // workaround for API sometimes failing on some Tokens(?)
      expect(error.message).toMatch(/Only chainId \d+ is supported/);
    }
  });

  test('Get_Markets', async () => {
    const markets = await paraSwap.swap.getAdapters();
    expect(markets.length).toBeGreaterThan(15);
  });

  test('Get_Tokens', async () => {
    const tokens = await paraSwap.swap.getTokens();

    expect(Array.isArray(tokens)).toBe(true);
    expect(tokens.length).toBeGreaterThan(0);
    expect(tokens[0]).toEqual(
      expect.objectContaining({
        symbol: expect.any(String),
        address: expect.any(String),
        decimals: expect.any(Number),
      })
    );
  });

  test('Get_Rates', async () => {
    const priceRoute = await paraSwap.swap.getRate({
      srcToken: ETH,
      destToken: DAI,
      amount: srcAmount,
      userAddress: senderAddress,
      side: SwapSide.SELL,
      options: {
        includeDEXS: ['UniswapV2'],
        otherExchangePrices: true,
      },
    });

    const { destAmount, bestRoute, others } = priceRoute;

    expect(typeof destAmount).toBe('string');

    expect(Array.isArray(bestRoute)).toBe(true);

    const swapExchange = bestRoute[0]?.swaps[0]?.swapExchanges[0];

    assert(swapExchange, 'exchange available at swapExchanges[0]');

    expect(typeof swapExchange.destAmount).toBe('string');
    expect(new BigNumber(swapExchange.destAmount).isNaN()).toBe(false);

    expect(typeof swapExchange.exchange).toBe('string');

    const firstBestRoute = bestRoute[0];
    assert(firstBestRoute, 'route available at bestRoute[0]');

    expect(typeof firstBestRoute.percent).toBe('number');
    expect(new BigNumber(firstBestRoute.percent).isNaN()).toBe(false);

    expect(typeof swapExchange.srcAmount).toBe('string');
    expect(new BigNumber(swapExchange.srcAmount).isNaN()).toBe(false);

    expect(Array.isArray(others)).toBe(true);

    const firstRoute = others?.[0];

    assert(firstRoute, 'at least one route must exist');

    expect(typeof firstRoute.exchange).toBe('string');

    expect(typeof firstRoute.unit).toBe('string');
    expect(firstRoute.unit && new BigNumber(firstRoute.unit).isNaN()).toBe(
      false
    );
  });

  test('Get_SwapTxData', async () => {
    const { priceRoute, txParams } = await paraSwap.swap.getSwapTxData({
      srcToken: ETH,
      destToken: DAI,
      amount: srcAmount,
      userAddress: senderAddress,
      side: SwapSide.SELL,
      slippage: 500,
      options: {
        includeDEXS: ['UniswapV2'],
      },
    });

    const bestRouteStable = priceRoute.bestRoute.map((b) => ({
      ...b,
      swaps: b.swaps.map((s) => ({
        ...s,
        swapExchanges: s.swapExchanges.map((se) => ({
          ...se,
          destAmount: 'dynamic_number',
          data: {
            ...se.data,
            gasUSD: 'dynamic_number',
          },
        })),
      })),
    }));

    const priceRouteStable = {
      ...priceRoute,
      gasCost: 'dynamic_number',
      gasCostUSD: 'dynamic_number',
      hmac: 'dynamic_number',
      destAmount: 'dynamic_number',
      blockNumber: 'dynamic_number',
      srcUSD: 'dynamic_number',
      destUSD: 'dynamic_number',
      bestRoute: bestRouteStable,
    };

    const txParamsStable = {
      ...txParams,
      data: 'dynamic_string',
      from: 'dynamic_string',
      gasPrice: 'dynamic_number',
    };

    expect(txParams.from).toEqual(senderAddress);

    expect(priceRouteStable).toMatchSnapshot('Get_SwapTxData::priceRoute');
    expect(txParamsStable).toMatchSnapshot('Get_SwapTxData::txParams');
  });

  test('Get_Spender', async () => {
    const spender = await paraSwap.swap.getSpender();
    expect(web3provider.utils.isAddress(spender));
  });

  test('Get_Allowance', async () => {
    const allowance = await paraSwap.swap.getAllowance(
      DUMMY_ADDRESS_FOR_TESTING_ALLOWANCES,
      DAI
    );

    assert(isAllowance(allowance), 'hardcoded dummy address should be found');

    expect(allowance.allowance).toEqual('123000000000000000');
  });

  test('Get_Adapters', async () => {
    const adapters = await paraSwap.swap.getAdapters();
    expect(adapters).toMatchSnapshot('Get_Adapters');
  });

  test('Build_Tx', async () => {
    const destToken = '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48';
    const priceRoute = await paraSwap.swap.getRate({
      srcToken,
      destToken,
      amount: srcAmount,
      userAddress: senderAddress,
      side: SwapSide.SELL,
      options: {
        includeDEXS: ['UniswapV2'],
      },
    });

    const destAmount = new BigNumber(priceRoute.destAmount)
      .times(0.99)
      .toFixed(0);

    const txParams = await paraSwap.swap.buildTx(
      {
        srcToken,
        destToken,
        srcAmount,
        destAmount,
        priceRoute,
        userAddress: senderAddress,
        partner: referrer,
      },
      { ignoreChecks: true }
    );

    expect(typeof txParams).toBe('object');
  });
  test('Build_and_Send_Tx', async () => {
    const priceRoute = await paraSwap.swap.getRate({
      srcToken,
      destToken,
      amount: srcAmount,
      userAddress: senderAddress,
      side: SwapSide.SELL,
      options: {
        includeDEXS: ['Uniswap', 'UniswapV2', 'Balancer', 'Oasis'],
      },
    });

    const destAmount = new BigNumber(priceRoute.destAmount)
      .times(0.99)
      .toFixed(0);

    const txParams = await paraSwap.swap.buildTx(
      {
        srcToken,
        destToken,
        srcAmount,
        destAmount,
        priceRoute,
        userAddress: signer.address,
        partner: referrer,
      },
      { ignoreChecks: true }
    );

    const transaction = {
      ...txParams,
      gasPrice:
        txParams.gasPrice &&
        '0x' + new BigNumber(txParams.gasPrice).toString(16),
      maxFeePerGas:
        txParams.maxFeePerGas &&
        '0x' + new BigNumber(txParams.maxFeePerGas).toString(16),
      maxPriorityFeePerGas:
        txParams.maxPriorityFeePerGas &&
        '0x' + new BigNumber(txParams.maxPriorityFeePerGas).toString(16),
      gasLimit: '0x' + new BigNumber(5000000).toString(16),
      value: '0x' + new BigNumber(txParams.value).toString(16),
    };
    const toContract = new ethers.Contract(destToken, erc20abi, ethersProvider);
    const beforeFromBalance = await ethersProvider.getBalance(signer.address);
    const beforeToBalance = await toContract.balanceOf(signer.address);

    const txr = await signer.sendTransaction(transaction);
    await txr.wait(1);
    const afterFromBalance = await ethersProvider.getBalance(signer.address);
    const afterToBalance = await toContract.balanceOf(signer.address);
    expect(beforeFromBalance.gt(afterFromBalance)).toBeTruthy();
    expect(beforeToBalance.lt(afterToBalance)).toBeTruthy();
  });
  test('Build_and_Send_Tx_BUY', async () => {
    const destAmount = srcAmount;
    const priceRoute = await paraSwap.swap.getRate({
      srcToken,
      destToken,
      amount: destAmount,
      userAddress: senderAddress,
      side: SwapSide.BUY,
      options: { includeDEXS: ['Uniswap', 'UniswapV2', 'Balancer', 'Oasis'] },
    });
    const _srcAmount = new BigNumber(priceRoute.srcAmount)
      .times(1.1)
      .toFixed(0);

    const txParams = await paraSwap.swap.buildTx(
      {
        srcToken,
        destToken,
        srcAmount: _srcAmount,
        destAmount,
        priceRoute,
        userAddress: signer.address,
        partner: referrer,
      },
      { ignoreChecks: true }
    );

    const transaction = {
      ...txParams,
      gasPrice:
        txParams.gasPrice &&
        '0x' + new BigNumber(txParams.gasPrice).toString(16),
      maxFeePerGas:
        txParams.maxFeePerGas &&
        '0x' + new BigNumber(txParams.maxFeePerGas).toString(16),
      maxPriorityFeePerGas:
        txParams.maxPriorityFeePerGas &&
        '0x' + new BigNumber(txParams.maxPriorityFeePerGas).toString(16),
      gasLimit: '0x' + new BigNumber(5000000).toString(16),
      value: '0x' + new BigNumber(txParams.value).toString(16),
    };
    const toContract = new ethers.Contract(destToken, erc20abi, ethersProvider);
    const beforeFromBalance = await ethersProvider.getBalance(signer.address);
    const beforeToBalance = await toContract.balanceOf(signer.address);

    const txr = await signer.sendTransaction(transaction);
    await txr.wait(1);
    const afterFromBalance = await ethersProvider.getBalance(signer.address);
    const afterToBalance = await toContract.balanceOf(signer.address);
    expect(beforeFromBalance.gt(afterFromBalance)).toBeTruthy();
    expect(beforeToBalance.lt(afterToBalance)).toBeTruthy();
  });
});

describe.each([
  [
    'fetch & ethersV5',
    { fetch },
    {
      ethersProviderOrSigner: signer,
      EthersContract: ethers.Contract,
      account: senderAddress,
    },
  ],
  [
    'fetch & ethersV6',
    { fetch },
    {
      ethersV6ProviderOrSigner: signerV6,
      EthersV6Contract: ethersV6.Contract,
      account: senderAddress,
    },
  ],
  ['axios & web3', { axios }, { web3: web3provider, account: senderAddress }],
])(
  'ParaSwap SDK: contract calling methods: %s',
  (testName, fetcherOptions, providerOptions) => {
    let paraSwap: SimpleSDK;

    beforeAll(() => {
      paraSwap = constructSimpleSDK(
        { chainId, ...fetcherOptions, version: '5' },
        providerOptions
      );
    });
    test('approveToken', async () => {
      const txHash = await paraSwap.swap.approveToken('12345', DAI);

      await ethersProvider.waitForTransaction(txHash);

      const toContract = new ethers.Contract(
        destToken,
        erc20abi,
        ethersProvider
      );
      const spender = await paraSwap.swap.getSpender();
      const allowance: BigNumberEthers = await toContract.allowance(
        signer.address,
        spender
      );
      expect(allowance.toString()).toEqual('12345');
    });
  }
);
