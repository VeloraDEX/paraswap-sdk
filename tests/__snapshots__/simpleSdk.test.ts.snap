// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ParaSwap SDK: fetcher made with: axios Get_Adapters: Get_Adapters 1`] = `
[
  "Bancor",
  "Compound",
  "DODOV1",
  "DODOV2",
  "TraderJoeV2.1",
  "Smoothy",
  "CurveV2",
  "Lido",
  "EtherFi",
  "CurveV1",
  "Swerve",
  "BalancerV1",
  "BalancerV2",
  "UniswapV2",
  "DefiSwap",
  "LinkSwap",
  "PancakeSwapV2",
  "SushiSwap",
  "ShibaSwap",
  "Verse",
  "UniswapV3",
  "SushiSwapV3",
  "PancakeswapV3",
  "AaveV2",
  "AaveV3",
  "KyberDmm",
  "Weth",
  "PolygonMigrator",
  "MakerPsm",
  "Synapse",
  "ParaSwapLimitOrders",
  "SolidlyV2",
  "Synthetix",
  "CurveV1Factory",
  "CurveV1StableNg",
  "wstETH",
  "Hashflow",
  "MaverickV1",
  "SwaapV2",
  "AngleStakedStableUSD",
  "AngleStakedStableEUR",
  "SolidlyV3",
  "Wombat",
  "Swell",
  "Spark",
  "sUSDS",
  "AugustusRFQ",
]
`;

exports[`ParaSwap SDK: fetcher made with: axios Get_SwapTxData: Get_SwapTxData::priceRoute 1`] = `
{
  "bestRoute": [
    {
      "percent": 100,
      "swaps": [
        {
          "destDecimals": 18,
          "destToken": "0x6b175474e89094c44da98b954eedeac495271d0f",
          "srcDecimals": 18,
          "srcToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
          "swapExchanges": [
            {
              "data": {
                "factory": "0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f",
                "feeFactor": 10000,
                "gasUSD": "dynamic_number",
                "initCode": "0x96e8ac4277198ff8b6f785478aa9a39f403cb768dd02cbee326c3e7da348845f",
                "path": [
                  "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                  "0x6b175474e89094c44da98b954eedeac495271d0f",
                ],
                "pools": [
                  {
                    "address": "0xA478c2975Ab1Ea89e8196811F51A7B7Ade33eB11",
                    "direction": false,
                    "fee": 30,
                  },
                ],
                "router": "0xF9234CB08edb93c0d4a4d4c70cC3FfD070e78e07",
              },
              "destAmount": "dynamic_number",
              "exchange": "UniswapV2",
              "percent": 100,
              "poolAddresses": [
                "0xA478c2975Ab1Ea89e8196811F51A7B7Ade33eB11",
              ],
              "srcAmount": "1000000000000000000",
            },
          ],
        },
      ],
    },
  ],
  "blockNumber": "dynamic_number",
  "contractAddress": "0xDEF171Fe48CF0115B1d80b88dc8eAB59176FEe57",
  "contractMethod": "swapOnUniswapV2Fork",
  "destAmount": "dynamic_number",
  "destDecimals": 18,
  "destToken": "0x6b175474e89094c44da98b954eedeac495271d0f",
  "destUSD": "dynamic_number",
  "gasCost": "dynamic_number",
  "gasCostUSD": "dynamic_number",
  "hmac": "dynamic_number",
  "maxImpactReached": false,
  "network": 1,
  "partner": "anon",
  "partnerFee": 0,
  "side": "SELL",
  "srcAmount": "1000000000000000000",
  "srcDecimals": 18,
  "srcToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
  "srcUSD": "dynamic_number",
  "tokenTransferProxy": "0x216b4b4ba9f3e719726886d34a177484278bfcae",
  "version": "5",
}
`;

exports[`ParaSwap SDK: fetcher made with: axios Get_SwapTxData: Get_SwapTxData::txParams 1`] = `
{
  "chainId": 1,
  "data": "dynamic_string",
  "from": "dynamic_string",
  "gasPrice": "dynamic_number",
  "to": "0xDEF171Fe48CF0115B1d80b88dc8eAB59176FEe57",
  "value": "1000000000000000000",
}
`;

exports[`ParaSwap SDK: fetcher made with: fetch Get_Adapters: Get_Adapters 1`] = `
[
  "Bancor",
  "Compound",
  "DODOV1",
  "DODOV2",
  "TraderJoeV2.1",
  "Smoothy",
  "CurveV2",
  "Lido",
  "EtherFi",
  "CurveV1",
  "Swerve",
  "BalancerV1",
  "BalancerV2",
  "UniswapV2",
  "DefiSwap",
  "LinkSwap",
  "PancakeSwapV2",
  "SushiSwap",
  "ShibaSwap",
  "Verse",
  "UniswapV3",
  "SushiSwapV3",
  "PancakeswapV3",
  "AaveV2",
  "AaveV3",
  "KyberDmm",
  "Weth",
  "PolygonMigrator",
  "MakerPsm",
  "Synapse",
  "ParaSwapLimitOrders",
  "SolidlyV2",
  "Synthetix",
  "CurveV1Factory",
  "CurveV1StableNg",
  "wstETH",
  "Hashflow",
  "MaverickV1",
  "SwaapV2",
  "AngleStakedStableUSD",
  "AngleStakedStableEUR",
  "SolidlyV3",
  "Wombat",
  "Swell",
  "Spark",
  "sUSDS",
  "AugustusRFQ",
]
`;

exports[`ParaSwap SDK: fetcher made with: fetch Get_SwapTxData: Get_SwapTxData::priceRoute 1`] = `
{
  "bestRoute": [
    {
      "percent": 100,
      "swaps": [
        {
          "destDecimals": 18,
          "destToken": "0x6b175474e89094c44da98b954eedeac495271d0f",
          "srcDecimals": 18,
          "srcToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
          "swapExchanges": [
            {
              "data": {
                "factory": "0x5C69bEe701ef814a2B6a3EDD4B1652CB9cc5aA6f",
                "feeFactor": 10000,
                "gasUSD": "dynamic_number",
                "initCode": "0x96e8ac4277198ff8b6f785478aa9a39f403cb768dd02cbee326c3e7da348845f",
                "path": [
                  "0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2",
                  "0x6b175474e89094c44da98b954eedeac495271d0f",
                ],
                "pools": [
                  {
                    "address": "0xA478c2975Ab1Ea89e8196811F51A7B7Ade33eB11",
                    "direction": false,
                    "fee": 30,
                  },
                ],
                "router": "0xF9234CB08edb93c0d4a4d4c70cC3FfD070e78e07",
              },
              "destAmount": "dynamic_number",
              "exchange": "UniswapV2",
              "percent": 100,
              "poolAddresses": [
                "0xA478c2975Ab1Ea89e8196811F51A7B7Ade33eB11",
              ],
              "srcAmount": "1000000000000000000",
            },
          ],
        },
      ],
    },
  ],
  "blockNumber": "dynamic_number",
  "contractAddress": "0xDEF171Fe48CF0115B1d80b88dc8eAB59176FEe57",
  "contractMethod": "swapOnUniswapV2Fork",
  "destAmount": "dynamic_number",
  "destDecimals": 18,
  "destToken": "0x6b175474e89094c44da98b954eedeac495271d0f",
  "destUSD": "dynamic_number",
  "gasCost": "dynamic_number",
  "gasCostUSD": "dynamic_number",
  "hmac": "dynamic_number",
  "maxImpactReached": false,
  "network": 1,
  "partner": "anon",
  "partnerFee": 0,
  "side": "SELL",
  "srcAmount": "1000000000000000000",
  "srcDecimals": 18,
  "srcToken": "0xeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeeee",
  "srcUSD": "dynamic_number",
  "tokenTransferProxy": "0x216b4b4ba9f3e719726886d34a177484278bfcae",
  "version": "5",
}
`;

exports[`ParaSwap SDK: fetcher made with: fetch Get_SwapTxData: Get_SwapTxData::txParams 1`] = `
{
  "chainId": 1,
  "data": "dynamic_string",
  "from": "dynamic_string",
  "gasPrice": "dynamic_number",
  "to": "0xDEF171Fe48CF0115B1d80b88dc8eAB59176FEe57",
  "value": "1000000000000000000",
}
`;
